gocd-server, gocd-agent: 19.3.0-8959 -> 21.2.0-12498

skeleten opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The current versions of `gocd-server` and `gocd-agent` are not only way out of date, but also do not properly function at this point in time. As such I felt the need to update the packages and services to be up to date and functional. This conists of a version bum p and some minor change to the start script used for the systemd-Units. I also made the java package used to execute the services configurable and set `adoptopenjdk-hotspot-bin-15` as the default, since the current version of `gocd-server` requires a specific java version of 15 or [crash](https://gist.github.com/skeleten/8cbe11f1a620af4e702abebd62b63c95).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

skeleten committed:
gocd-agent: 19.3.0-8959 -> 21.2.0-12498

skeleten committed:
gocd-server: 19.3.0-8959 -> 21.2.0-12498

skeleten committed:
gocd-agent: Update service to respect moved file

The file `go-agent/agent-bootstrapper.jar` moved to
`go-agent/lib/agent-bootstrapper.jar` with the bump to gocd-agent
version 21.2.0-12489

skeleten committed:
gocd-agent, gocd-server: Make java package an option and set working default

tilpner reviewed (commented):
Congratulations on your first PR, thanks for sticking with it! :)

Hashes match [upstream](https://www.gocd.org/download/#zip), `gocd-agent` and `gocd-server` build successfully here, module not tested.

I'm not sure about putting java into the service path, added a few questions about that.

tilpner commented on nixos/modules/services/continuous-integration/gocd-agent/default.nix:
```diff
@@ -35,9 +35,18 @@ in {
       packages = mkOption {
-        default = [ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ];
-        defaultText = "[ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ]";
+        default = [ pkgs.stdenv pkgs.adoptopenjdk-hotspot-bin-15 pkgs.git config.programs.ssh.package pkgs.nix ];
+        defaultText = "[ pkgs.stdenv pkgs.adoptopenjdk-hotspot-bin-15 pkgs.git config.programs.ssh.package pkgs.nix ]";
```
Duplicating the list as a string seems really error-prone, there's got to be a better way.

skeleten commented:
Hey,
thanks a lot for your feedback. I've changed the packet to `adoptopenjdk-jre-hotspot-bin-15` and removed it from the `PATH`.
These changes did not seem to have any affect on the execution of these binaries and I agree these changes clean up the service a bit :)

r-rmcgibbo commented:
Result of `nixpkgs-review pr 133901` at 0f6e63a5 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>gocd-agent</li>
    <li>gocd-server</li>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/name-and-version.md">name-and-version</a>

Did you mean to pass `pname` instead of `name` to `mkDerivation`?

Near pkgs/development/tools/continuous-integration/gocd-agent/default.nix:4:3:
```
  |
4 |   name = "gocd-agent-${version}-${rev}";
  |   ^
```

Near pkgs/development/tools/continuous-integration/gocd-agent/default.nix:5:3:
```
  |
5 |   version = "21.2.0";
  |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/name-and-version.md">name-and-version</a>

Did you mean to pass `pname` instead of `name` to `mkDerivation`?

Near pkgs/development/tools/continuous-integration/gocd-server/default.nix:4:3:
```
  |
4 |   name = "gocd-server-${version}-${rev}";
  |   ^
```

Near pkgs/development/tools/continuous-integration/gocd-server/default.nix:5:3:
```
  |
5 |   version = "21.2.0";
  |   ^
```
</li>
  </ul>
</details>

tilpner commented on nixos/modules/services/continuous-integration/gocd-server/default.nix:
```diff
@@ -67,9 +67,18 @@ in {
-        default = [ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ];
-        defaultText = "[ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ]";
+        default = [ pkgs.stdenv pkgs.git config.programs.ssh.package pkgs.nix ];
+        defaultText = "[ pkgs.stdenv pkgs.git config.programs.ssh.package pkgs.nix ]";
```
Very minor nitpick: I believe these should be using `literalExample`, so that they can be rendered without quotes

tilpner reviewed (approved):
Issues have been addressed, but I also couldn't figure out how to deduplicate defaultText.

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/gocd-server/default.nix:
```diff
@@ -2,12 +2,12 @@
   src = fetchurl {
     url = "https://download.go.cd/binaries/${version}-${rev}/generic/go-server-${version}-${rev}.zip";
-    sha256 = "0c30qzd6awlw0zx91rk6na0mmgykqkgrw9ychx18ivjwma0hr0sc";
+    sha256 = "a465d23c55bd853636af6cff277b85665054fbd5864a8f13b48b2979f04ccac6";
```
In the future please use base encoded hashes.

skeleten replied:
I was generating them using `sha256sum`, which seemed to work. I have since then learned that `nix-prefetch-url` is the preferred way to generate these hashes. Is replacing these with the generated ones from `nix-prefetch-url` okay?

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/gocd-server/default.nix:
```diff
@@ -2,12 +2,12 @@
   name = "gocd-server-${version}-${rev}";
-  version = "19.3.0";
-  rev = "8959";
+  version = "21.2.0";
+  rev = "12498";
```
```suggestion
  version = "21.2.0-12498";
```
then we could also convert name to pname and simplify the download URL.

skeleten replied:
Thanks for your feedback! I mostly left the existing code as is, which is why I decided to leave the version and name structure. While your suggested changes simplify the URL indeed, they would also require a little bit more complexity (that is, splitting the String back up again) in the build script. Is that acceptable?

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/gocd-agent/default.nix:
```diff
@@ -2,12 +2,12 @@
-  version = "19.3.0";
-  rev = "8959";
+  version = "21.2.0";
+  rev = "12498";
 
   src = fetchurl {
     url = "https://download.go.cd/binaries/${version}-${rev}/generic/go-agent-${version}-${rev}.zip";
-    sha256 = "1nirdv82i8x4s1dyb0rmxldh8avappd4g3mbbl6xp7r7s0drcprp";
+    sha256 = "105e38058cb918bf0ff2c8a2458a911fddf2ff347d8e9ba7c9107972d58439f2";
```
Why are we repeating this?

skeleten replied:
I am honestly not sure what exactly you're referring to here; Both the URL as well as the hash differ from the `gocd-server` ones; While I'm assuming the same feedback as with that file still applies here, its not a repetition as far as I can tell

skeleten committed:
gocd-agent, gocd-server: refactor

This implements feedback given on the related PR:

 - Combine `version` and `rev`
 - Use `pname` rather than `name`
 - Use the proper hashes acquired by `nix-prefetch-url`

skeleten committed:
nixos/gocd-agent, nixos/gocd-server: Make use of `literalExample`

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/gocd-server/default.nix:
```diff
@@ -20,9 +19,14 @@ stdenv.mkDerivation rec {
+    in
+    ''
+      unzip $src -d $out
+      mv $out/go-server-${baseVersion} $out/go-server
```
Can we use a wildcard ``*`` instead? It could simplify this section by a lot.

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/gocd-agent/default.nix:
```diff
@@ -19,8 +19,13 @@ stdenv.mkDerivation rec {
+    in
+    ''
+      unzip $src -d $out
+      mv $out/go-agent-${baseVersion} $out/go-agent
```
Can we use a wildcard ``*`` instead? It could simplify this section by a lot.

SuperSandro2000 commented on nixos/modules/services/continuous-integration/gocd-server/default.nix:
```diff
@@ -67,9 +67,18 @@ in {
-        default = [ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ];
-        defaultText = "[ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ]";
+        default = [ pkgs.stdenv pkgs.git config.programs.ssh.package pkgs.nix ];
+        defaultText = literalExample "[ pkgs.stdenv pkgs.git config.programs.ssh.package pkgs.nix ]";
```
Isn't stdenv and nix available everywhere by default?

SuperSandro2000 commented on nixos/modules/services/continuous-integration/gocd-agent/default.nix:
```diff
@@ -35,9 +35,18 @@ in {
-        default = [ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ];
-        defaultText = "[ pkgs.stdenv pkgs.jre pkgs.git config.programs.ssh.package pkgs.nix ]";
+        default = [ pkgs.stdenv pkgs.git config.programs.ssh.package pkgs.nix ];
+        defaultText = literalExample "[ pkgs.stdenv pkgs.git config.programs.ssh.package pkgs.nix ]";
```
Isn't stdenv and nix available everywhere by default?

skeleten commented:
I'm closing this PR as it is stale and outdated
